File: sentimental-credit/credit.py
import re
from cs50 import get_string


def validator_input(input: str) -> bool:
    validator = r"^\d+$"
    if (re.match(validator, input)):
        return True
    return False


def sum_total(number_card: str) -> int:
    change = False
    sum_total = 0
    for num in number_card[::-1]:
        num = int(num)
        if (change):
            num *= 2
            num = (num//10) + (num % 10)
        sum_total += num
        change = not change
    return sum_total


def select_card(number_card: str) -> None:
    size = len(number_card)
    first_digits = int(number_card[0:2])

    if ((size == 15) and (first_digits == 34 or first_digits == 37)):
        print("AMEX")
    elif ((size == 16 or size == 13) and (first_digits // 10) == 4):
        print("VISA")
    elif ((size == 16) and 55 >= first_digits >= 51):
        print("MASTERCARD")
    else:
        print("INVALID")


if (__name__ == "__main__"):
    while True:
        number = get_string("Number: ")
        if (validator_input(number)):
            break

    if (sum_total(number) % 10 == 0):
        select_card(number)
    else:
        print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

long sum_total(long num);
void select_card(long num);
int two_first_digits(long num);
int count_num(long num);

int main(void)
{
    long num_card = get_long("Numbers: ");

    if (sum_total(num_card) % 10 == 0)
    {
        select_card(num_card);
    }
    else
    {
        printf("INVALID\n");
    }
}

void select_card(long num)
{
    int size = count_num(num);
    int digits = two_first_digits(num);

    if ((size == 15) && (digits == 34 || digits == 37))
    {
        printf("AMEX\n");
    }
    else if ((size == 16 || size == 13) && (digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else if ((size == 16) && (digits >= 51 && digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int two_first_digits(long num)
{
    while (num >= 100)
    {
        num /= 10;
    }
    return num;
}

int count_num(long num)
{
    int size = 0;
    while (num > 0)
    {
        size++;
        num /= 10;
    }
    return size;
}

long sum_total(long num)
{
    int assistant = 0;
    int sum = 0;

    while (num > 0)
    {
        int digit = num % 10;
        if (assistant)
        {
            digit *= 2;
            sum += (digit / 10) + (digit % 10);
        }
        else
        {
            sum += digit;
        }
        num /= 10;
        assistant = !assistant;
    }
    return sum;
}
